Feature missing reference support code

seabbs opened the pull request:
This PR adds support code for adding modelling of missing reference dates (#43). This includes updating `epinowcast()` to have `enw_missing()` as an argument (it cannot be optional as stan complains about missing data even when it has zero dimensions), updating `enw_missing()` to have a not on mode, adding support for missing effects and priors, adding a new grouped version of the observation likelihood in which we can add likelihood changes from #107, and adding a prototype function for simulating missing data (we should probably look at adding this and `enw_incidence_to_cumulative()` to the package but I am aware we are getting lots of exported functions and it may be a bit overwhelming/hard to support in the long term.

Alongside #107 this leaves post-processing as a final support step in terms of supporting missing dates to the same level as non-missing dates. 

*Note the grouped observation likelihood is likely not quite there - tricky keeping track of the indexing. It would be easier if both grouped and snapshot versions could be used without missing data (as should return the same thing) but this might be a bit annoying to support internally.*

adrian-lison reviewed (approved):
Again, great work, really close now to getting the missingness model running.

If I understand correctly, for the missingness likelihood, we
- still need to integrate the ref_miss_prop
- add a matrix storing the without-reference date obs while going through the snapshots
- do the broadcasting as proposed in #107 for the without-reference date matrix
- transform the matrix into a vector by log-sum-exping over the columns and
- add another call to obs_lmpfs for this second vector

Anything else?

seabbs commented on inst/examples/germany_missing.R:
```diff
@@ -0,0 +1,92 @@
+  by_with_group_id <- c(".group", by)
+  obs <- enw_new_reports(obs)
+
+  obs[, missing := floor(new_confirm * proportion)]
```
yeah, this is just the super crude version. I wasn't sure what we wanted in terms of simulators. Ideally, we would have something to simulate the whole model but that requires a lot of code. I think the stochastic version would be nicer. Will implement here.

seabbs replied:
Done in https://github.com/epiforecasts/epinowcast/pull/138/commits/77707bc458b0a6e2cafe9335ae05bdfdda63d069

seabbs commented on inst/stan/epinowcast.stan:
```diff
@@ -22,78 +22,128 @@ data {
-  int refp_rncol; // number of standard deviations to use for pooling
-  matrix[refp_fncol, refp_rncol + 1] refp_rdesign; // Pooling pmf design matrix 
-  int model_refp; // parametric distribution (0 = none, 1 = exp, 2 = lognormal, 3 = gamma, 4 = loglogistic)
+  int model_refp;
+  int refp_fnrow;
+  array[s] int refp_findex;
+  int refp_fncol;
+  matrix[refp_fnrow, refp_fncol + 1] refp_fdesign;
+  int refp_rncol;
+  matrix[refp_fncol, refp_rncol + 1] refp_rdesign; 
+  array[2] real refp_mean_int_p;
+  array[2] real refp_sd_int_p;
+  array[2] real refp_mean_beta_sd_p;
+  array[2] real refp_sd_beta_sd_p; 
```
So I felt like it was repeating known information as these parts come from a function with documented output and it was the same every time. The idea of removing was to make the stan code easier/clearer to read just because there was less of it. Perhaps a halfway house is documenting what each thing is once and then saying it is the same for all other prefixes

adrian-lison replied:
I understand the rationale

seabbs commented on inst/stan/epinowcast.stan:
```diff
@@ -22,78 +22,128 @@ data {
+  int miss_fnindex;
+  int miss_fncol;
+  int miss_rncol;
+  matrix[miss_fnindex, miss_fncol + 1] miss_fdesign;
```
We don't have it here as we don't need it because `miss_` is not sparse and so no lookup between rows and `_fnindex` is needed. That definitely needs to be better/at least a bit documented. We could have it regardless but it just wouldn't be used.

adrian-lison replied:
Okay, I think it is better to not have it here to avoid overhead, but it would be good if we could add a comment about this somewhere. Maybe just a "sparse specification" / "dense specification" comment at the beginning of each block?

seabbs replied:
I've made an issue for this (#153) and suggest we deal with it in another PR with additional doc changes in the R code to make the general approach as clear as possible.

seabbs commented on inst/stan/epinowcast.stan:
```diff
@@ -22,78 +22,128 @@ data {
   int ref_as_p = (model_rep > 0 || model_refp > 0) ? 0 : 1; 
+  for (i in 1:g) {
+    groups[i] = i;
+  }
```
ooo nice!

seabbs commented on inst/stan/epinowcast.stan:
```diff
@@ -22,78 +22,128 @@ data {
-  vector<lower=0>[rep_rncol] rep_beta_sd; // pooled modifiers to report date
-  array[model_obs > 0 ? 1 : 0] real<lower=0> sqrt_phi; // Overall dispersion by group
+
+  // Reference model
+  array[model_refp ? 1 : 0] real<lower=-10, upper=logdmax> refp_mean_int;
+  array[model_refp > 1 ? 1 : 0]real<lower=1e-3, upper=2*dmax> refp_sd_int; 
+  vector[model_refp ? refp_fncol : 0] refp_mean_beta; 
+  vector[model_refp > 1 ? refp_fncol : 0] refp_sd_beta; 
+  vector<lower=0>[refp_rncol] refp_mean_beta_sd;
+  vector<lower=0>[model_refp ? refp_rncol : 0] refp_sd_beta_sd; 
+
+  // Report model
+  vector[rep_fncol] rep_beta;
+  vector<lower=0>[rep_rncol] rep_beta_sd; 
+
+  // Missing reference date model
+  array[model_miss] real miss_int;
+  vector[miss_fncol] miss_beta; 
+  vector<lower=0>[miss_rncol] miss_beta_sd; 
+
+  // Observation model
+  array[model_obs > 0 ? 1 : 0] real<lower=0> sqrt_phi; // Overall dispersion
```
I wanted to be able to avoid forcing an intercept but ultimately couldn't. Maybe something for the future if its something people actually want control of (in most cases I think this as is likely fine)

seabbs commented on inst/stan/epinowcast.stan:
```diff
@@ -22,78 +22,128 @@ data {
-  array[model_obs > 0 ? 1 : 0] real phi; // Transformed overdispersion (joint across all observations)
-  // calculate log mean and sd parameters for each dataset from design matrices
+  // Missing model
+  vector[miss_fnindex] miss_ref_prop;
```
yes - or we can do it the other way around. I thought it would be easier to model low proportions vs high (though maybe it is exactly the same).

seabbs replied:
note in https://github.com/epiforecasts/epinowcast/pull/138/commits/706388c909b38dfc465b6a7f4f3b47ff1c64eff3 I updated this to be on the log scale for stability and consistency.

adrian-lison replied:
No it's good, actually miss_ref may also be more intuitive

seabbs commented:
Yes, I think you have got all the changes we need. We also need some postprocessing work to make plotting/summary etc part of the tooling. The only other thing I want to add to this PR is some tests that the two `delay_` functions return the same thing and potentially expose the ability to switch between them to the user.

seabbs commented:
Note that rather than adding tests I instead add more lower level functions. I also used these to update the generated quantities to use the flat data structure.

seabbs commented:
> still need to integrate the ref_miss_prop

Added in https://github.com/epiforecasts/epinowcast/pull/138/commits/706388c909b38dfc465b6a7f4f3b47ff1c64eff3

seabbs commented:
I think everything planned for this PR is now in place so pinged for another review. As quite a few `png` updates in this PR I have opted for a squash merge.

A quick summary of what is here (sorry got a little out of hand): 

- Passes in missing model input to stan code
- Added model machinery for the missingness model effects to stan
- Updates handling of `enw_missing()` to make it work with stan
- Reorganises the stan code slightly to be easier to read, including dropping repeated comments
- Splitting `delay_lmpf` into two functions one that is aggregated over snapshots and one over groups. Expose this to the user as a choice and forcing the use of the grouped option when the missing model is present.
- Both of these functions now depend on a vectorised version of `expected_obs_from_index` called `expected_obs_from_snaps`. 
- This has also been used to refactor the generated quantities. I have also switched all parts of this to use the flattened data storage structure. 
- Added regression tests for posterior predictions, nowcasts and key parameters. These are rough and only a first pass but better than nothing
- Added a new example for missing data along with some prototype functions for making missing data (these likely can be refined and added to the package in another PR).

seabbs commented:
Woops!

adrian-lison commented on inst/stan/epinowcast.stan:
```diff
@@ -68,31 +91,63 @@ transformed data{
   // if no reporting day effects use native probability for reference day
   // effects, i.e. do not convert to logit hazard
   int ref_as_p = (model_rep > 0 || model_refp == 0) ? 0 : 1; 
+  // Type of likelihood aggregation to use
+  int ll_aggregation = likelihood_aggregation + model_miss;
```
I wonder whether it would be better to complain to the user instead of just switching to group aggregation in case a missingness model is used... With the current approach, the user is likely unaware that a different parallelization scheme is used. Maybe we should at least warn the user (see my commit)

seabbs replied:
Yeah good point. Part of me thinks this should just work for the user as it's a complexity they really don't need to know but I see your point of we allow it to be set then it should work as they expect with warnings.

seabbs replied:
Really like your proposed check function. Seems like a good solution!

adrian-lison reviewed (approved):
Looks all good to me. I added one proposal commit
